pallets/foreign-investments/src/lib.rs

```diff
@@ -43,43 +43,131 @@
 
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use cfg_types::investments::Swap;
-/// Edit this file to define custom logic or remove it if it is not needed.
-/// Learn more about FRAME and the core library of Substrate FRAME pallets:
-/// <https://docs.substrate.io/reference/frame-pallets/>
+use cfg_traits::{investments::TrancheCurrency, PoolInspect};
+use cfg_types::investments::{CollectedAmount, Swap};
+use frame_support::{dispatch::DispatchResult, ensure};
 pub use pallet::*;
-
-pub mod errors;
-pub mod hooks;
-pub mod impls;
-pub mod types;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{traits::EnsureSub, ArithmeticError, DispatchError};
 
 #[cfg(test)]
 mod mock;
 
 #[cfg(test)]
 mod tests;
 
+/// Hold the base information of a foreign investment/redemption
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct BaseInfo<T: Config> {
+	foreign_currency: T::CurrencyId,
+	pool_currency: T::CurrencyId,
+	collected_amount: CollectedAmount<T::Balance>,
+}
+
+impl<T: Config> BaseInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			foreign_currency,
+			pool_currency: T::PoolInspect::currency_for(investment_id.of_pool())
+				.ok_or(Error::<T>::PoolNotFound)?,
+			collected_amount: CollectedAmount::default(),
+		})
+	}
+
+	fn ensure_same_foreign(&self, foreign_currency: T::CurrencyId) -> DispatchResult {
+		ensure!(
+			self.foreign_currency == foreign_currency,
+			Error::<T>::MismatchedForeignCurrency
+		);
+
+		Ok(())
+	}
+}
+
+/// Hold the information of a foreign investment
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct InvestmentInfo<T: Config> {
+	base: BaseInfo<T>,
+	total_foreign_amount: T::Balance,
+}
+
+impl<T: Config> InvestmentInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			base: BaseInfo::new(investment_id, foreign_currency)?,
+			total_foreign_amount: T::Balance::default(),
+		})
+	}
+
+	fn remaining_foreign_amount(&self) -> Result<T::Balance, ArithmeticError> {
+		self.total_foreign_amount
+			.ensure_sub(self.base.collected_amount.amount_collected)
+	}
```

Comment:
> For investments, the amount_collected is the amount of collected tranche tokens. What we need here is amount_payment converted into foreign, because amount_payment is the amount of pool currency.

You're right! I was almost sure I had some error with those quantities. Thanks for your eyes!

> Thus, for any (foreign, pool) pair which does not swap 1:1, we can't determine the foreign amount from the pool amount upfront synchronously and deterministically 

Uff... you're right, and I do not like it this issue.

Just a thought. Why do not specify the `amount_remaining` as `pool_currency` instead? I mean, it made sense to specify it in foreign currencies before, but now, with market prices, we should specify what makes sense for the chain and is more "immutable," in this case, pool currencies.